clang/lib/Sema/SemaOverload.cpp

```diff
@@ -960,18 +960,13 @@ static bool shouldAddReversedEqEq(Sema &S, SourceLocation OpLoc,
     return true;
   }
   // Otherwise the search scope is the namespace scope of which F is a member.
-  LookupResult NonMembers(S, NotEqOp, OpLoc,
-                          Sema::LookupNameKind::LookupOperatorName);
-  S.LookupName(NonMembers,
-               S.getScopeForContext(EqFD->getEnclosingNamespaceContext()));
-  NonMembers.suppressAccessDiagnostics();
-  for (NamedDecl *Op : NonMembers) {
-    auto *FD = Op->getAsFunction();
-    if(auto* UD = dyn_cast<UsingShadowDecl>(Op))
-      FD = UD->getUnderlyingDecl()->getAsFunction();
-    if (FunctionsCorrespond(S.Context, EqFD, FD) &&
-        declaresSameEntity(cast<Decl>(EqFD->getDeclContext()),
-                           cast<Decl>(Op->getDeclContext())))
+  for (NamedDecl *Op : EqFD->getEnclosingNamespaceContext()->lookup(NotEqOp)) {
+    auto *NotEqFD = Op->getAsFunction();
```

Comment:
What about template functions? Something like:
```
namespace ns {
template <class T>
struct A {
};

template <class T> 
struct B : A<T> {
};

template <class T> bool operator == (B<T>, A<T>);
template <class T> bool operator != (B<T>, A<T>);
}

void test() {
    ns::A<int> a;
    ns::B<int> b;
    a == b;
}
```
https://gcc.godbolt.org/z/sc36a7f81

Reply:
Added this test. This fails now as expected.

Reply:
Ah, getAsFunction picks a primary template. LG and thanks for adding a test!